tasks/gettext-checker.js

```diff
@@ -52,31 +56,64 @@ module.exports = function (grunt) {
-            grunt.log.error(keyDiff);
-            grunt.fail.fatal('Check FAILED.');
+        var keyDiff;
+
+        if (options.checkPoKeys) {
+            // find items in template.pot that are not in this po file
+            keyDiff = _.difference(potKeys, poKeys);
+            if (keyDiff.length > 0 ) {
+                grunt.log.error('The following translation keys in ' + options.potFile + ' are not present in ' + options.poFile);
+                grunt.log.error(keyDiff);
+                grunt.fail.fatal('Check FAILED.');
+            } else {
+                grunt.log.ok('All keys from .pot template are present in .po file.');
+            }
         }
 
-        // find items in this po file which are not in template.pot
-        var keyDiff = _.difference(poKeys, potKeys);
-        if (keyDiff.length > 0 ) {
-            grunt.log.error('The following translation keys in ' + options.poFile + ' are not present in template.pot');
-            grunt.log.error(keyDiff);
-            grunt.fail.fatal('Check FAILED.');
+        if (options.checkPotKeys) {
+            // find items in this po file which are not in template.pot
+            keyDiff = _.difference(poKeys, potKeys);
+            if (keyDiff.length > 0 ) {
+                grunt.log.error('The following translation keys in ' + options.poFile + ' are not present in ' + options.potFile);
+                grunt.log.error(keyDiff);
+                grunt.fail.fatal('Check FAILED.');
+            } else {
+                grunt.log.ok('All keys from .po file are present in .pot template.');
+            }
         }
 
-        grunt.log.ok('Translation .pot template and all .po file keys match.')
+        if (options.checkKeyOrder) {
+            // Check if keys in .pot file and .po file are in same order. We use _.find to drop out as soon as we find one non-matching
+            // to make things a bit faster if we can
+            var matchingOrder = _.find(potKeys, function (item, index) {
+                if (poKeys[index] !== item) {
+                    // The keys don't match, so are out of order
+                    return true;
+                }
+                return false;
+            });
+            // _.find returns undefined if all keys are in matching order, so we error if it is not undefined
+            if (!_.isUndefined(matchingOrder)) {
+                grunt.log.error(
+                    'The keys in ' + options.potFile + ' and ' + options.poFile + ' are not in the same order.\r\n' +
+                    'If you were not expecting the order to change, please check your generating tool version and ' +
+                    'environment to check it matches the environment used by the project.\r\n' +
+                    'If you have deliberately changed the order of the .pot file, you must update the order of the ' +
+                    '.po files as well - e.g. Using a tool such as poEdit\`s "Update from template" option.'
+                );
```

Comment:
Ah I see!

On 9 March 2015 at 10:34, Graham Gilchrist <notifications@github.com> wrote:

> In tasks/gettext-checker.js
> <https://github.com/incuna/grunt-gettext-checker/pull/3#discussion_r26025669>
> :
>
> > +            var matchingOrder = _.find(potKeys, function (item, index) {
> > +                if (poKeys[index] !== item) {
> > +                    // The keys don't match, so are out of order
> > +                    return true;
> > +                }
> > +                return false;
> > +            });
> > +            // _.find returns undefined if all keys are in matching order, so we error if it is not undefined
> > +            if (!_.isUndefined(matchingOrder)) {
> > +                grunt.log.error(
> > +                    'The keys in ' + options.potFile + ' and ' + options.poFile + ' are not in the same order.\r\n' +
> > +                    'If you were not expecting the order to change, please check your generating tool version and ' +
> > +                    'environment to check it matches the environment used by the project.\r\n' +
> > +                    'If you have deliberately changed the order of the .pot file, you must update the order of the ' +
> > +                    '.po files as well - e.g. Using a tool such as poEdit\`s "Update from template" option.'
> > +                );
>
> The newlines are for specific separate sentences, rather than to wrap at
> 80 chars, but we should use grunt.log.errorlns(msg) anyway to wrap the
> rest
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/incuna/grunt-gettext-checker/pull/3/files#r26025669>.
>



-- 
Henry Blyth - Developer
henry@incuna.com
01865 791 494
Incuna Ltd, 16 Middle Way, Oxford, OX2 7LH